diff --git a/plugins/selection/EntityReference_SelectionHandler_Generic.class.php b/plugins/selection/EntityReference_SelectionHandler_Generic.class.php
index 57a3a37..4ff4dcd 100644
--- a/plugins/selection/EntityReference_SelectionHandler_Generic.class.php
+++ b/plugins/selection/EntityReference_SelectionHandler_Generic.class.php
@@ -251,15 +251,20 @@ class EntityReference_SelectionHandler_Generic implements EntityReference_Select
       }
     }
 
-    // Add a generic entity access tag to the query.
-    $query->addTag($this->field['settings']['target_type'] . '_access');
+    if (!empty($this->field['settings']['handler_settings']['sort'])) {
+      $sort_settings = $this->field['settings']['handler_settings']['sort'];
+    }
+
+    if (empty($sort_settings) || $sort_settings['type'] != 'field') {
+      // Add a generic entity access tag to the query.
+      $query->addTag($this->field['settings']['target_type'] . '_access');
+    }
     $query->addTag('entityreference');
     $query->addMetaData('field', $this->field);
     $query->addMetaData('entityreference_selection_handler', $this);
 
     // Add the sort option.
-    if (!empty($this->field['settings']['handler_settings']['sort'])) {
-      $sort_settings = $this->field['settings']['handler_settings']['sort'];
+    if (!empty($sort_settings)) {
       if ($sort_settings['type'] == 'property') {
         $query->propertyOrderBy($sort_settings['property'], $sort_settings['direction']);
       }
