diff --git a/includes/handlers.inc b/includes/handlers.inc
index 4346133..bc8e141 100644
--- a/includes/handlers.inc
+++ b/includes/handlers.inc
@@ -1090,14 +1090,12 @@ function views_break_phrase_string($str, &$handler = NULL) {
   }
 
   // Determine if the string has 'or' operators (plus signs) or 'and' operators
-  // (commas) and split the string accordingly. If we have an 'and' operator,
-  // spaces are treated as part of the word being split, but otherwise they are
-  // treated the same as a plus sign.
-  $or_wildcard = '[^\s+,]';
+  // (commas) and split the string accordingly.
+  $or_wildcard = '[^+,]';
   $and_wildcard = '[^+,]';
-  if (preg_match("/^({$or_wildcard}+[+ ])+{$or_wildcard}+$/", $str)) {
+  if (preg_match("/^({$or_wildcard}+[+])+{$or_wildcard}+$/", $str)) {
     $handler->operator = 'or';
-    $handler->value = preg_split('/[+ ]/', $str);
+    $handler->value = preg_split('/[+]/', $str);
   }
   elseif (preg_match("/^({$and_wildcard}+,)*{$and_wildcard}+$/", $str)) {
     $handler->operator = 'and';
